README.md
# flutter_cs_cache

Flutter package to provide simple and extendible cache feature to be used in flutter applications.

## Getting Started
Import CsCache as 

```import 'package:flutter_cs_cache/flutter_cs_cache.dart';```

Once imported make sure you have created CsCache instance ahead of time, before using getKey or setKey functions

### To set a key value
csCache.setKey(key: keyName, value: valueString, expireOn: expiryTimeInMillisecondsSinceEpoch);

### To get a key value
csCache.getKey(key: keyName);

### To remove a key value
csCache.removeKey(key: keyName);

### To purge all keys which might have been expired
csCache.purgeCache();

### To clear all cache entrie (expired and non-expired)
csCache.clearCache();


## example code

```
class _MyHomePageState extends State<MyHomePage> {
  String cacheMessage = "";
  final CsCache cache = new CsCache();
  final String _cacheKey = "1111";

  void _checkCache() {
    String cacheEntry = cache.getKey(key: _cacheKey);
    if(cacheEntry == null) cache.setKey(key: _cacheKey, value: DateTime.now().toString(), expireOn: DateTime.now().add(new Duration(seconds: 300)).millisecondsSinceEpoch);
    setState(() {
      cacheMessage = cacheEntry ?? "Cache entry not found";
    });
  }

  void _removeCacheEntry(){
    cache.removeKey(key: _cacheKey);
    _checkCache();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text(
              'Cache has following value:',
              style: Theme.of(context).textTheme.title,
            ),
            new Text(
              '$cacheMessage',
              style: Theme.of(context).textTheme.body1,
            ),

            SizedBox(height: 50.0,),
            RaisedButton(
              child: Text("Remove Cache Entry"),
              onPressed: _removeCacheEntry,
              )
          ],
        ),
      ),
      floatingActionButton: new FloatingActionButton(
        onPressed: _checkCache,
        tooltip: 'Add cache entry',
        child: new Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```


CHANGELOG.md
## 0.0.1 - Initial Release

* Basic cache features included such as set/get/remove/purge/clear cache


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

// import 'package:flutter/material.dart';
// import 'package:flutter_test/flutter_test.dart';

// import 'package:example/main.dart';

void main() {
  // testWidgets('Counter increments smoke test', (WidgetTester tester) async {
  //   // Build our app and trigger a frame.
  //   await tester.pumpWidget(new MyApp());

  //   // Verify that our counter starts at 0.
  //   expect(find.text('0'), findsOneWidget);
  //   expect(find.text('1'), findsNothing);

  //   // Tap the '+' icon and trigger a frame.
  //   await tester.tap(find.byIcon(Icons.add));
  //   await tester.pump();

  //   // Verify that our counter has incremented.
  //   expect(find.text('0'), findsNothing);
  //   expect(find.text('1'), findsOneWidget);
  // });
}


test/flutter_cs_cache_test.dart
import 'package:test/test.dart';

import 'package:flutter_cs_cache/flutter_cs_cache.dart';

void main() {
  test('Test cache entry is null', () {
    final CsCache csCache = new CsCache();
    final String cacheValue = csCache.getKey(key: "randomkey-${DateTime.now().millisecondsSinceEpoch}"  );
    expect(cacheValue, null);
  });
}


lib/flutter_cs_cache.dart
library flutter_cs_cache;

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// A Calculator.
class CsCache {
  static CsCache _singltonInstance;
  SharedPreferences _sharedPreferences;

  factory CsCache() {
    if (_singltonInstance == null) {
      _singltonInstance = CsCache._internal();
      _singltonInstance._setSharedPref();
    }

    return _singltonInstance;
  }

  CsCache._internal();

  void _setSharedPref() async {
    _sharedPreferences = await SharedPreferences.getInstance();
  }

  bool isValid(List<String> cacheValueEntry) {
    try {
      if (cacheValueEntry != null &&
          cacheValueEntry.isNotEmpty &&
          cacheValueEntry.length == 2) {
        int expiryTimestamp = int.parse(cacheValueEntry[0]);
        return (expiryTimestamp >= DateTime.now().millisecondsSinceEpoch);
      }
    } catch (e) {
      print(e.toString());
    }
    return false;
  }

  void setKey(
      {@required String key, @required String value, int expireOn = -1}) {
    List<String> cacheValueEntry = [expireOn.toString(), value];

    _sharedPreferences.setStringList(key, cacheValueEntry);
  }

  String getKey({@required String key}) {
    List<String> cacheValueEntry = _sharedPreferences.getStringList(key);
    if (isValid(cacheValueEntry)) {
      return cacheValueEntry[1];
    }

    return null;
  }

  void removeKey({@required String key}) {
    _sharedPreferences.remove(key);
  }

  void purgeCache() {
    Set<String> keys = _sharedPreferences.getKeys();
    for (String key in keys) {
      if (getKey(key: key) == null) {
        _sharedPreferences.remove(key);
      }
    }
  }

  void clearCache() {
    _sharedPreferences.clear();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_cs_cache/flutter_cs_cache.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Cache Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String cacheMessage = "";
  final CsCache cache = new CsCache();
  final String _cacheKey = "1111";

  void _checkCache() {
    String cacheEntry = cache.getKey(key: _cacheKey);
    if(cacheEntry == null) cache.setKey(key: _cacheKey, value: DateTime.now().toString(), expireOn: DateTime.now().add(new Duration(seconds: 300)).millisecondsSinceEpoch);
    setState(() {
      cacheMessage = cacheEntry ?? "Cache entry not found";
    });
  }

  void _removeCacheEntry(){
    cache.removeKey(key: _cacheKey);
    _checkCache();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text(
              'Cache has following value:',
              style: Theme.of(context).textTheme.title,
            ),
            new Text(
              '$cacheMessage',
              style: Theme.of(context).textTheme.body1,
            ),

            SizedBox(height: 50.0,),
            RaisedButton(
              child: Text("Remove Cache Entry"),
              onPressed: _removeCacheEntry,
              )
          ],
        ),
      ),
      floatingActionButton: new FloatingActionButton(
        onPressed: _checkCache,
        tooltip: 'Add cache entry',
        child: new Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
